ltq-vdsl-app: extend ubus with dsl statistics

turboproc opened the pull request:
This pull requests introduces an extension the the ubus dsl integration with a command to 
expose the DSL statistics data required to render the dsl charts. Statistics include bit-allocation,
SNR, QLN and HLOG data.

With this extension `ubus` calls will look like
```
root:~# ubus -v list dsl
'dsl' @88f73b6b
        "metrics":{}
        "statistics":{}
```

Output of `ubus call dsl statistics` looks like (trailer):
```
{
        "bits": {
                "downstream": {
                        "groups": 4096,
                        "data": [
                                0,
                                0,
                                0,
                                0,
                                0,
                                0,
```
This total output is a kind of long but follows similar structure for upstream as for downstream and continues
with SNR, QLN and HLog data.

turboproc committed:
ltq-vdsl-app: extend ubus call to provide DSL statistics

Adding a new method to `ubus call dsl` to retrieve DSL statistics
used to feed the DSL charts (bit allocation, SNR, QLN and HLOG)

Signed-off-by: Roland Barenbrug <roland@treslong.com>

dhewg commented on package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c:
```diff
@@ -837,6 +948,7 @@ static int metrics(struct ubus_context *ctx, struct ubus_object *obj,
 
 static const struct ubus_method dsl_methods[] = {
 	UBUS_METHOD_NOARG("metrics", metrics),
+	UBUS_METHOD_NOARG("statistics", statistics)
```
That's very generic, `line_statistics` maybe?

dhewg reviewed (commented):
Nice, thanks for properly adding this info here!

dhewg commented:
Now that I mentioned dB, the current `m_db` function doesn't take invalid values into account.
Do you mind fixing that with an additional commit for this PR?

Something like
```
static inline void m_db(const char *id, int value, int invalid) {
	if (value <> invalid)
		m_double(id, (double)value / 10);
}
```

and then
```
	// invalid value indicator taken from drv_dsl_cpe_api_g997.h
	m_db("latn", out.data.LATN, 1271);
	m_db("satn", out.data.SATN, 1271);
	m_db("snr", out.data.SNR, -641);
	m_db("actps", out.data.ACTPS, -901);
	m_db("actatp", out.data.ACTATP, -512);
```

turboproc commented:
@dhewg, thnx for your feedback. Will work through item by item. Generally, it is probably more convenient to do the translation into real values in this part rather than presenting an internal representation. Will also simplify the graph rendering part.

turboproc commented on package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c:
```diff
@@ -146,7 +146,7 @@ static inline void m_bool(const char *id, bool value) {
 
 static inline void m_u32(const char *id, uint32_t value) {
-	blobmsg_add_u32(&b, id, value);
+	blobmsg_add_u64(&b, id, value);
```
Clear, will check

dhewg replied:
Still needs a rebase

turboproc commented on package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c:
```diff
@@ -157,10 +157,20 @@ static inline void m_db(const char *id, int value) {
+}
+
+
+static inline void m_array_dlt_16(const char *id, const uint16_t *value, uint16_t len) {
```
Especially when doing the value translation here, negative values require a 16 bit space hence ` m_array_16` is still required.

dhewg replied:
Related to the comments below: Just as with m_db I meant to provide dB float values. Then this may not be required anymore

turboproc commented on package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c:
```diff
@@ -724,6 +765,76 @@ static void describe_mode(standard_t standard, profile_t profile, vector_t vecto
+                   struct blob_attr *msg)
+{
+        int fd, fd_mei;
+        void *c, *c2;
```
Clear

turboproc commented on package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c:
```diff
@@ -724,6 +765,76 @@ static void describe_mode(standard_t standard, profile_t profile, vector_t vecto
+	blobmsg_close_table(&b, c2);
+	blobmsg_close_table(&b, c);
+
+		
```
Clear (do we have somewhere the applicable formatting rules?, May be even a formatting checker?).

dhewg replied:
It's more like every package has it's own ;)
As a rule of thumb: keep the one you see in the file you're working on

turboproc replied:
Done

turboproc commented on package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c:
```diff
@@ -724,6 +765,76 @@ static void describe_mode(standard_t standard, profile_t profile, vector_t vecto
+		
+	c = blobmsg_open_table(&b, "bits");
+	c2 = blobmsg_open_table(&b, "downstream");
+	g977_get_bit_allocation(fd,DSL_DOWNSTREAM);
```
Done

turboproc commented on package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c:
```diff
@@ -413,6 +423,37 @@ static void g997_line_inventory(int fd) {
+	IOCTL_DIR_DELT(DSL_G997_DeltSnr_t, DSL_FIO_G997_DELT_SNR_GET, direction, DSL_DELT_DATA_SHOWTIME);
+	m_u32("groupsize", out.data.nGroupSize);
+	m_u32("groups", out.data.deltSnr.nNumData);
+	m_array("data", &out.data.deltSnr.nNSCData, out.data.deltSnr.nNumData);
```
Done, so we're doing the translation to real values in the program. Still requires a value to indicate an error value. 127 seems to work for all cases.

dhewg replied:
For ubus values here, we only provide valid values. If something isn't available or invalid we just don't provide it. A missing value is the indicator for that. That matches expected behavior when plotting the metrics from here in e.g. grafana.

dhewg replied:
Wait, you're dealing with arrays here. Skipping a value means you don't know the original index of a value. Is that okay or do you require that info?

turboproc replied:
> Wait, you're dealing with arrays here. Skipping a value means you don't know the original index of a value. Is that okay or do you require that info?

That's exactly the point. Dealing with array's which means that at any index you want to provide a valid value or an indication that no valid value is available.

dhewg replied:
How about NAN then?
For m_db skipping invalid ones still seems best though

dhewg replied:
That looks better, but I meant the actual `NAN` define/constant from math.h instead of a string. Does that work?

turboproc commented on package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c:
```diff
@@ -413,6 +423,37 @@ static void g997_line_inventory(int fd) {
+	IOCTL_DIR_DELT(DSL_G997_DeltQln_t, DSL_FIO_G997_DELT_QLN_GET, direction, DSL_DELT_DATA_SHOWTIME);
+	m_u32("groupsize", out.data.nGroupSize);
+	m_u32("groups", out.data.deltQln.nNumData);
+	m_array("data", &out.data.deltQln.nNSCData, out.data.deltQln.nNumData);
```
Same, had to use int16 values as range starts at -150.

turboproc replied:
Switched at this point to float as well.

turboproc commented on package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c:
```diff
@@ -413,6 +423,37 @@ static void g997_line_inventory(int fd) {
+	IOCTL_DIR_DELT(DSL_G997_DeltHlog_t, DSL_FIO_G997_DELT_HLOG_GET, direction, DSL_DELT_DATA_SHOWTIME);
+	m_u32("groupsize", out.data.nGroupSize);
+	m_u32("groups", out.data.deltHlog.nNumData);
+	m_array_dlt_16("data", &out.data.deltHlog.nNSCData, out.data.deltHlog.nNumData);
```
Using HLog (not Hlin). Again had to use int16 values as the negative values didn't come out (appeared as positive values).

turboproc replied:
Switched in the mean time to float (double seems a bit overdone but I might be wrong on that point).

dhewg replied:
I didn't verify, but you may be right. We don't need that precision I guess? I don't recall if there was another reason to use double vs float. If you choose float, please change m_db as well to keep it consistent

dhewg replied:
Seems because there's only BLOBMSG_TYPE_DOUBLE :)

turboproc replied:
Will do. Have it essentially all working apart of the other change you asked for in the existing code. Might commit something new tonight.

turboproc commented:
> Now that I mentioned dB, the current `m_db` function doesn't take invalid values into account. Do you mind fixing that with an additional commit for this PR?
> 
> Something like
> 
> ```
> static inline void m_db(const char *id, int value, int invalid) {
> 	if (value != invalid)
> 		m_double(id, (double)value / 10);
> }
> ```
> 
> and then
> 
> ```
> 	// invalid value indicators taken from drv_dsl_cpe_api_g997.h
> 	m_db("latn", out.data.LATN, 1271);
> 	m_db("satn", out.data.SATN, 1271);
> 	m_db("snr", out.data.SNR, -641);
> 	m_db("actps", out.data.ACTPS, -901);
> 	m_db("actatp", out.data.ACTATP, -512);
> ```

Need to cook on this. Still need a mechanism to indicate error values.

dhewg commented:
As mentioned above: The snipped I posted intentionally skips invalid values. That's the indicator, there's no need for special values

turboproc committed:
ltq-vdsl-app: changes and fixes

Introducing a number of changes and fixes since previous commit:
- transformation to real values rather than internal representation
- in arrays a NaN will be used in case a value is not available
- latn, satn, snr, actps and actatp only provided if value available
- fixed indenting and spaces

Signed-off-by: Roland Barenbrug <roland@treslong.com>

janh commented:
I think this should include a version bump for `ltq-adsl-app`, as that package copies `dsl_cpe_ubus.c` from `ltq-vdsl-app`.

turboproc committed:
ltq-vdsl-app: define special value as border case

Although not require by relevant standards (IT-T recommendation)
the 0 value has been defined as a corner case indicating essentially
that data is not available. Apparently some firmware versions do
not strictly follow the standards.

Signed-off-by: Roland Barenbrug <roland@treslong.com>

dhewg commented on package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c:
```diff
@@ -413,6 +415,70 @@ static void g997_line_inventory(int fd) {
+		if (out.data.deltSnr.nNSCData[i] != 255 && out.data.deltSnr.nNSCData[i] != 255 != 0) {
+			m_double("", -32 + (double)out.data.deltSnr.nNSCData[i] / 2); // SNR -32 ... 95 dB
+		} else {
+			m_str("", "NaN");
```
github sucks, from the other discussion marked as resolved:
That looks better, but I meant the actual NAN define/constant from math.h instead of a string. Does that work?

turboproc replied:
On the NaN topic, that might not be trivial. Doing m_double("", NAN), results in `nan` being output in the json file. That would be OK but `nan` is considered by json as a string and should appear in between "" hence resulting in invalid json. If m_double("", NAN) would result in `"null"` as output that would at least result in valid json. Conclusion: for now not sure whether it adds any value to change.

dhewg replied:
Alright, thanks for checking!
What I'm trying to avoid is to have every consumer having to special case whatever we special case here ;)
So if there's something for ubus/json we can use to represent these invalid values, we should use it.
So I guess `NaN` is out of the picture, in fact it's not part of the json spec and there's a [whole python issue](https://github.com/python/cpython/issues/84813) about it.

Stemming from that discussion is json's `null`. Can you please try that instead? It looks like it's represented by libubus' `BLOBMSG_TYPE_UNSPEC`, which we can use in `m_double`. Something like
```
static inline void m_double(const char *id, double value) {
	if (value = NAN)
		blobmsg_add_field(&b, BLOBMSG_TYPE_UNSPEC, id, NULL, 0);
	else
		blobmsg_add_double(&b, id, value);
}
```

And then just `m_double(id, NAN)` for those values. The json should be a valid `null` then.

dhewg replied:
The snipped above doesn't work, but it works on my branch I mentioned below

dhewg commented on package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c:
```diff
@@ -146,21 +146,23 @@ static inline void m_bool(const char *id, bool value) {
 
 static inline void m_u32(const char *id, uint32_t value) {
-	blobmsg_add_u32(&b, id, value);
+	blobmsg_add_u64(&b, id, value);
```
This isn't resolved, still needs a rebase

turboproc replied:
On the rebasing I could use a little help. Last time I tried this it ended up in a kind of mess.

dhewg replied:
You can continue with the cleanup and fixes I did on my branch.
Assuming you're on the branch from this PR, commit and then backup your changes in a new branch:
`git branch bak-dsl-ubus-stats`

And then switch over:
```
git remote add dhewg https://github.com/dhewg/openwrt.git
git fetch dhewg
git reset --hard dhewg/dsl-ubus-stat
```

dhewg commented on package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c:
```diff
@@ -413,6 +415,70 @@ static void g997_line_inventory(int fd) {
+		if (out.data.deltQln.nNSCData[i] != 255 && out.data.deltQln.nNSCData[i] != 0) {
+			m_double("", -23 - (double)out.data.deltQln.nNSCData[i] / 2); // QLN -150 ... -23 dBm/Hz
+		} else {
+			m_str("", "NaN");
```
same here

dhewg commented on package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c:
```diff
@@ -413,6 +415,70 @@ static void g997_line_inventory(int fd) {
+		if (out.data.deltHlog.nNSCData[i] != 1023 && out.data.deltHlog.nNSCData[i] != 0) {
+			m_double("", 6 - (double)out.data.deltHlog.nNSCData[i] / 10); // HLOG +6 ... -96 dB
+		} else {
+			m_str("", "NaN");
```
and here

dhewg commented on package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c:
```diff
@@ -413,6 +415,70 @@ static void g997_line_inventory(int fd) {
+	void *c = blobmsg_open_array(&b, "data");
+	
+	for (uint16_t i  = 0  ; i < out.data.deltSnr.nNumData ; i++) { //uint16_t i = 0; i < len; ++i
+		if (out.data.deltSnr.nNSCData[i] != 255 && out.data.deltSnr.nNSCData[i] != 255 != 0) {
```
copypaste fail `!= 255 != 0`

turboproc replied:
Indeed, will be fixed in next commit

dhewg commented:
I took a stab at getting invalid values to output as `null`. This works for me and includes various cleanups and fixes too:
https://github.com/dhewg/openwrt/commits/dsl-ubus-stats

turboproc committed:
ltq-vdsl-app: fix handling of NaN entries in json output

This commit fixes the handling of entries which essentially have no
value by using in that case a 'none' value. This seems to be the
cleanest way forward. Requires change in the luci module to handle
'none' values as a special case.

Signed-off-by: Roland Barenbrug <roland@treslng.com>

turboproc commented:
@feckert , fixed the NAN issue by making a change in m_double. In case of a NAN value it will introduce a "none" value. Having it in m_double makes it as transparent as possible.

turboproc committed:
ltq-vdsl-app: further fix to handle NaN in json output as null. The
JSON value 'null' (without quotes) is the JSON way to denote no value.

Again requires change in luci module to recognize 'null' as special
value

Signed-off-by: Roland Barenbrug <roland@treslong.com>

turboproc committed:
ltq-vdsl-app: further fix to handle NaN in json output as null. The
JSON value 'null' (without quotes) is the JSON way to denote no value.

Again requires change in luci module to recognize 'null' as special
value

Signed-off-by: Roland Barenbrug <roland@treslong.com>

turboproc committed:
Merge branch 'dsl-ubus-stats' of https://github.com/turboproc/openwrt into dsl-ubus-stats

turboproc committed:
ltq-vdsl-app: JSON output uses 'null' value in case of NaN

Following now JSON syntax and using 'null' value (without quotes) in cases
where there is no data to provide.

Further cleaned up some typos.

Signed-off-by: Roland Barenbrug <roland@treslong.com>

feckert commented:
@turboproc Something is wrong with the pullrequest, unfortunately I can't test it!
Please make one or more commits to your change.
Because it will go into the repository when it is merged.
One commit for each logical change , so that it is comprehensible what you have done.

dhewg commented:
As mentioned, I already splitted the commits logically and even fixed a few issues, see above. I even provided instructions on how to just use those, any issue with just doing that?

feckert commented:
@dhewg but @turboproc pushed also a merge commit https://github.com/openwrt/openwrt/pull/10420/commits this must be changed before merge

dhewg commented:
Indeed, which is why I also posted https://github.com/openwrt/openwrt/pull/10420#discussion_r962563806
That'll magically get rid of all problems ;)

feckert commented:
I would suggest we merge this now so we can move forward.
If something doesn't fit, we can still change it until the next release.
For me it currently looks like this.
Ubus output:
[statistics.txt](https://github.com/openwrt/openwrt/files/9573220/statistics.txt)
Luci view:
![Screenshot 2022-09-15 at 09-57-59 VR2-103954 - DSL line spectrum - LuCI](https://user-images.githubusercontent.com/553091/190350256-dafbace4-dff7-4082-a109-f2f866f2dec4.png)

turboproc commented:
> I would suggest we merge this now so we can move forward. If something doesn't fit, we can still change it until the next release. For me it currently looks like this. Ubus output: [statistics.txt](https://github.com/openwrt/openwrt/files/9573220/statistics.txt) Luci view: ![Screenshot 2022-09-15 at 09-57-59 VR2-103954 - DSL line spectrum - LuCI](https://user-images.githubusercontent.com/553091/190350256-dafbace4-dff7-4082-a109-f2f866f2dec4.png)

Sounds like a plan. For now I wouldn't know what to improve further. Obviously you used the updated Luci module.

dhewg commented:
I honestly don't know what the issue is... you don't even have to fix it yourself, I already did.
I fixed a pointer error, multiple whitespace issues and got rid of the unnecessary mei fd.
There's no work involved, just use it and push it?

feckert commented:
@hauke could you please merge this so we could get on and so get more testers on the master branch.

hauke commented:
The version from this pull request has a merge commit in it. Please remove the merge commit. 
Do you prefer this version https://github.com/dhewg/openwrt/commits/dsl-ubus-stats ?

hauke commented on package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c:
```diff
@@ -146,21 +151,23 @@ static inline void m_bool(const char *id, bool value) {
 	void *c = blobmsg_open_array(&b, id);
 
-	for (uint8_t i = 0; i < len; ++i)
+	for (uint16_t i = 0; i < len; ++i)
```
When you already change it I would prefer `size_t` over `uint16_t` for both types. With size_t we should not run into an overflow.

feckert commented:
@turboproc package was renamed to ltq-vdsl-vr9-app could you please rebase your changes

turboproc commented:
> @turboproc package was renamed to ltq-vdsl-vr9-app could you please rebase your changes
> 
> Edit: And merge this into one commit `ltq-vdsl-app: extend ubus call to provide DSL statistics`

Sure, let me check this weekend. Kind of busy these days.

dhewg commented:
To prevent this from bitrot, let's please just merge my clean up version at https://github.com/dhewg/openwrt/commits/dsl-ubus-stats

@turboproc already refered that version:
https://github.com/openwrt/luci/pull/5895#issuecomment-1279768668

And it's tested on VR9 too, see the follow up comments there.

feckert commented:
Either @turboproc updates the pullrequest with the suggested changes from @dhewg or @dhewg opens a new one.
Otherwise we won't get anywhere here.
Only when this has been merged we can also merge the luci app.

dhewg commented:
Agreed in general, but please go ahead and open new PRs if that's what it takes :)
It's just annoying when review comments get ignored, and I feel I've invested enough time on this.

feckert commented:
> Agreed in general, but please go ahead and open new PRs if that's what it takes :) It's just annoying when review comments get ignored, and I feel I've invested enough time on this.

I can understand your point. But we were almost done.

For the feature:
@dhewg Is it really that much work to open a new pullrequest with the two changes from you?
@turboproc Couldn`t you have update your pullrequest with the two chnnges from @dhewg?

Since I would like to see the feature in OpenWrt, I took the feature from @turboproc and the changes @dhewg and created a new [pullrequest](https://github.com/openwrt/openwrt/pull/11088)

Anyway thank you for your work :+1:

turboproc commented:
> > Agreed in general, but please go ahead and open new PRs if that's what it takes :) It's just annoying when review comments get ignored, and I feel I've invested enough time on this.
> 
> I can understand your point. But we were almost done.
> 
> For the feature: @dhewg Is it really that much work to open a new pullrequest with the two changes from you? @turboproc Couldn`t you have update your pullrequest with the two changes from @dhewg?
> 
> Since I would like to see the feature in OpenWrt, I took the feature from @turboproc and the changes @dhewg and created a new [pullrequest](https://github.com/openwrt/openwrt/pull/11088)
> 
> Anyway thank you for your work 👍

@feckert , sorry got confused on who was supposed to make the last action. Thnx for taking this one,

dhewg commented:
> @dhewg Is it really that much work to open a new pullrequest with the two changes from you?

If I open a PR, and need to be motivated to in-cooperate feedback, or we're possibly in the same situation as now. So... ;)
Anyway, thanks for taking care of it, looking forward to seeing it finally getting merged!

hauke commented:
Thank you for your patches, I applied the versions from #11088 to master in 5787e0c9fe0ae218f745ea5dcf30e20cc028842e and cc5d8ae42731b50469ab92ed00be3cfc83d4ce11.
